### PR TITLE
2.9 to develop

### PR DESCRIPTION
**Pull request by SimonRichardson**
Conflicts!

```sh
CONFLICT (content): Merge conflict in version/version.go
CONFLICT (content): Merge conflict in upgrades/model_test.go
CONFLICT (content): Merge conflict in upgrades/model.go
CONFLICT (content): Merge conflict in snap/snapcraft.yaml
CONFLICT (content): Merge conflict in scripts/win-installer/setup.iss
CONFLICT (content): Merge conflict in migration/precheck_test.go
CONFLICT (content): Merge conflict in go.sum
CONFLICT (content): Merge conflict in go.mod
CONFLICT (content): Merge conflict in core/charm/repository.go
CONFLICT (content): Merge conflict in apiserver/facades/client/charms/charmstorerepo.go
CONFLICT (content): Merge conflict in apiserver/facades/client/charms/charmhubrepo.go
CONFLICT (content): Merge conflict in api/charms/client_test.go
CONFLICT (content): Merge conflict in api/charms/client.go
CONFLICT (content): Merge conflict in api/application/client_test.go
```

381f3b15ce (upstream/2.9, origin/2.9, 2.9) Merge pull request #12695 from SimonRichardson/list-charm-resources
86d8b6abbb Merge pull request #12691 from SimonRichardson/add-coverage
95b785f35e Merge pull request #12699 from wallyworld/merge-2.8-20210226
38f89aaafc Merge pull request #12685 from hmlanigan/deploy-bundle-test
78ba9eee8b Merge pull request #12690 from achilleasa/2.9-normalize-output-messages-for-remove-cloud
450b5f02aa Merge pull request #12683 from SimonRichardson/charm-resources
ce1efe0cc8 Merge pull request #12687 from jujubot/increment-to-2.9-rc7
c163e9ec35 Merge pull request #12686 from wallyworld/merge-2.9-20210225
52d1547901 Merge pull request #12684 from achilleasa/2.9-improve-remove-cloud-output
8283c390e5 Merge pull request #12682 from SimonRichardson/charmstore-user-agent-version
24650ffa47 Merge pull request #12672 from hmlanigan/charmhub-download
d35f384329 Merge pull request #12680 from SimonRichardson/application-already-exists
9ee7b0381c Merge pull request #12658 from SimonRichardson/charmhub-bundle-channel
902bfd36d4 Merge pull request #12676 from tlm/merge-2.8-to-2.9
8cdfb00f6b Merge pull request #12678 from wallyworld/forward-port-12677
3eb499c155 Merge pull request #12675 from tlm/model-operator-lp1915320

### CODE REVIEW

**Comment by SimonRichardson** [comment]
!!build!!

File: api/application/client_test.go [hunk 1]
```diff
@@ -132,6 +132,71 @@ func (s *applicationSuite) TestDeploy(c *gc.C) {
...
+	})
+	args := application.DeployArgs{
+		NumUnits:      1,
+		AttachStorage: []string{"data/0"},
+	}
+	err := client.Deploy(args)
+	c.Assert(err, gc.ErrorMatches, "this juju controller does not support AttachStorage")
+	c.Assert(called, jc.IsFalse)
+}
+
```
Comment (wallyworld) [inline api/application/client_test.go:199]: These methods no longer exist in develop.

File: api/application/client.go [hunk 2]
```diff
@@ -162,7 +163,20 @@ func (c *Client) Deploy(args DeployArgs) error {
...
+	if pErr, ok := errors.Cause(err).(*params.Error); ok {
+		switch pErr.Code {
+		case params.CodeAlreadyExists:
+			// Remove the "already exists" in the error message to prevent
+			// stuttering.
+			msg := strings.Replace(err.Error(), " already exists", "", -1)
+			return errors.AlreadyExistsf(msg)
+		}
+	}
+	return errors.Trace(err)
```
Comment (wallyworld) [inline api/application/client.go:179]: This code is not in develop

**Review by wallyworld — CHANGES_REQUESTED** [review]
There's a bunch of stuff that has been copied across from 2.9 that has already been deleted in develop branch so should not have been ported across

File: provider/azure/config.go [hunk 3]
```diff
@@ -99,16 +98,6 @@ type azureModelConfig struct {
...
-// knownStorageAccountTypes returns a list of valid storage SKU names.
-//
-// The term "account type" is is used in previous versions of the Azure SDK.
-func knownStorageAccountTypes() (accountTypes []string) {
-	for _, name := range storage.PossibleSkuNameValues() {
-		accountTypes = append(accountTypes, string(name))
-	}
-	return accountTypes
-}
-
```
Comment (wallyworld) [inline provider/azure/config.go:111]: Thank you, I should have deleted this

File: apiserver/facades/client/action/action_test.go [hunk 4]
```diff
@@ -242,33 +242,6 @@ type receiverGroup struct {
...
-		}, {
-			Receiver: names.NewUnitTag("mysql/0"),
-			Actions: []testCaseAction{
-				{"fakeaction", map[string]interface{}{"zxcv": false}, false},
-				{"fakeaction", map[string]interface{}{}, true},
-			},
-		},
-	},
-}}
-
```
Comment (wallyworld) [inline apiserver/facades/client/action/action_test.go:271]: Thank you

**Review by wallyworld — APPROVED** [review]
Looks good once the obsolete code in api/application is removed

File: api/application/client.go (hunk 2 above)
Comment (SimonRichardson) [reply api/application/client.go:179]: I added this, so we can better present an error message.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by petevg** [comment]
Smash merging this. Simon to investigate why the job is failing at the very end.

**Pull request merged**